clutel/clutel.go

```diff
@@ -31,3 +32,99 @@ func AddToOTELHTTPLabeler(
 		node.AddToOTELHTTPLabeler(ctx),
 	))
 }
+
+// GetSpan retrieves the current OpenTelemetry span from the context.
+func GetSpan(ctx context.Context) trace.Span {
+	return trace.SpanFromContext(ctx)
+}
+
+type spanBuilder struct {
+	kvs  map[string]any
+	opts []trace.SpanStartOption
+}
+
+// NewSpan produces a span builder that allows complete configuration and
+// attribution of the span before it gets started.
+func NewSpan() *spanBuilder {
+	return &spanBuilder{}
+}
+
+// WithAttrs adds attrs to the span.
+func (sb *spanBuilder) WithAttrs(kvs ...any) *spanBuilder {
+	if sb.kvs == nil {
+		sb.kvs = make(map[string]any)
+	}
+
+	// nc.AddValues(stringify.Normalize(kvs...))
+	sb.kvs = stringify.Normalize(kvs...)
+
+	return sb
+}
+
+// WithOpts configures the span with specific otel options.  Note that one
+// of the options can be `trace.WithAttributes()`; don't use that, use
+// spanBuilder.WithAttrs() instead.
+func (sb *spanBuilder) WithOpts(opts ...trace.SpanStartOption) *spanBuilder {
+	if sb.opts == nil {
+		sb.opts = make([]trace.SpanStartOption, 0, len(opts))
+	}
+
+	sb.opts = append(sb.opts, opts...)
+
+	return sb
+}
+
+// Start begins the span with the provided name and attaches it to the context.
+func (sb *spanBuilder) Start(
+	ctx context.Context,
+	name string,
+) context.Context {
+	nc := node.FromCtx(ctx)
+	ctx, spanned := nc.AddSpan(ctx, name, sb.opts...)
```

Comment:
this is missing a check if the current node is `nil`. IIRC, there was some race condition (see #151) that I was chasing related to that. It'd probably be best to have this function's logic be as close to the original `StartSpan` as possible

Reply:
I've added the check anyway, just for security.

Reply:
I think the issue wasn't that `nc` could be nil, but that not initializing OTEL or something similar led to repeatedly modifying the existing `nc` or something, leading to a test failure when `-race` was provided